tracer/events.go

```diff
@@ -178,8 +179,36 @@ func startPollingPerfEventMonitor(ctx context.Context, perfEventMap *ebpf.Map,
 					noDataCount.Add(1)
 					continue
 				}
-				triggerFunc(data.RawSample, data.CPU)
+
+				// Keep track of min KTime seen in this batch processing loop
+				trace := t.loadBpfTrace(data.RawSample, data.CPU)
+				if minKTime == 0 || trace.KTime < minKTime {
+					minKTime = trace.KTime
+				}
+				traceOutChan <- trace
+			}
+			// After we've received and processed all trace events,
+			// call SymbolizationComplete if there is a pending oldKTime
+			// that we haven't yet propagated to the rest of the agent.
+			//
+			// This introduces both an upper bound to SymbolizationComplete
+			// call frequency (dictated by pollTicker) but also skips calls
+			// when none are needed (e.g. no trace events have been read).
+			//
+			// We use oldKTime instead of minKTime (except when the latter is
+			// smaller than the former) to take into account scheduling delays
+			// that could in theory result in observed KTime going back in time.
+			if oldKTime > 0 {
+				kt := oldKTime
+				if minKTime > 0 && minKTime < kt {
+					// If current minKTime is smaller than oldKTime, use it
+					// instead of oldKTime (and set it to 0 to avoid a repeat).
+					kt = minKTime
+					minKTime = 0
+				}
+				t.TraceProcessor().SymbolizationComplete(kt)
 			}
+			oldKTime = minKTime
```

Comment:
Instead of calling `SymbolizationComplete()` for every trace in the scope of `HandleTrace()` it should be fine to call it every `pollTicker` intervall, after all traces were fetched.

The following could simplify significantly this complexity:
```suggestion

				traceOutChan <- t.loadBpfTrace(data.RawSample, data.CPU)
			}

			t.TraceProcessor().SymbolizationComplete(kt)		
```

With the proposed approached, I see two issues:

(a) After the first time fetching all events from the perf event buffer, `SymbolizationComplete` is not called, as `oldKTime` is `0`. For every other iteration of fetching events, `SymbolizationComplete()` will be called. If the idea is to always call `SymbolizationComplete()` except for the very first time, then we should do it in a different way.

(b) Around line 165 `minKTime = 0` is a new variable for each iteration of `PollLoop` . So setting `minKTime = 0` around line 205 will not have an effect, as for the next iteration, there will be a new `minKTime`.

Reply:
I'm assuming our chat today addressed this. For other reviewers, not calling `SymbolizationComplete` when `oldKTime==0` is by design as it allows us to introduce a small delay that will minimize the probability of receiving a trace from the kernel with `KTime` less than what we previously reported via `SymbolizationComplete`.

Essentially, we're always reporting the previous poll iteration's `minKTime` rather than the current `minKTime`.